src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs

```diff
@@ -140,53 +142,56 @@ public partial class NetworkChange
                 throw new NetworkInformationException(message);
             }
 
-            s_socket = new SocketWrapper(newSocket);
+            s_socket = new Socket(new SafeSocketHandle(newSocket, ownsHandle: true));
 
-            new Thread(args => LoopReadSocket((SocketWrapper)args!))
-            {
-                IsBackground = true,
-                Name = ".NET Network Address Change"
-            }.UnsafeStart(s_socket);
+            // Don't capture ExecutionContext.
+            ThreadPool.UnsafeQueueUserWorkItem(static socket =>
+                _ = ReadEventsAsync(socket),
+                s_socket, preferLocal: false);
         }
 
         private static void CloseSocket()
         {
             Debug.Assert(s_socket != null, "s_socket was null when CloseSocket was called.");
-            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket != null ? s_socket.Socket : -1);
-            if (result != Interop.Error.SUCCESS)
-            {
-                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
-                throw new NetworkInformationException(message);
-            }
-
+            s_socket.Dispose();
             s_socket = null;
         }
 
-        private static unsafe void LoopReadSocket(SocketWrapper initiallyCreatedSocket)
+        private static async Task ReadEventsAsync(Socket socket)
         {
-            while (s_socket == initiallyCreatedSocket) //if both references are equal then s_socket was not changed
+            while (s_socket == socket)
             {
-                //we can continue processing events
-                Interop.Sys.ReadEvents(initiallyCreatedSocket.Socket, &ProcessEvent);
+                // Wait for data to become available.
+                await socket.ReceiveAsync(Array.Empty<byte>(), SocketFlags.None).ConfigureAwait(false);
+
+                if (!TryReadEvents(socket))
+                {
+                    return; // Stop reading on error/EOF.
+                }
             }
+
+            static unsafe bool TryReadEvents(Socket socket)
+                => Interop.Sys.ReadEvents(socket.SafeHandle, &ProcessEvent) >= 0;
         }
 
         [UnmanagedCallersOnly]
-        private static void ProcessEvent(int socket, Interop.Sys.NetworkChangeKind kind)
+        private static void ProcessEvent(IntPtr socket, Interop.Sys.NetworkChangeKind kind)
         {
             if (kind != Interop.Sys.NetworkChangeKind.None)
             {
                 lock (s_gate)
                 {
-                    if (s_socket != null && socket == s_socket.Socket)
+                    // It's safe to compare raw handle values because ProcessEvents gets
+                    // called from ReadEvents which holds a reference on the SafeHandle.
+                    if (s_socket != null && socket == s_socket.Handle)
                     {
                         OnSocketEvent(kind);
```

Comment:
Preexisting your change, but it's a little concerning we're invoking arbitrary user callbacks while holding the s_gate lock.

Reply:
For good or bad, this causes the events to synchronize with the handler registrations.

Reply:
Ugh. Ok.